### PR TITLE
Delay force refresh by DefaultInterval when OCI GetNodePool call retu…

### PR DESCRIPTION
**Pull request by vbhargav875**
What type of PR is this?

/kind feature

What this PR does / why we need it:

In the OCI Nodepools Implementation of ClusterAutoscaler, we do not set any delay to the calls that ClusterAutoscaler makes to the OKE GetNodePool API when the API returns a 404 error (which is possible when the Nodepool is deleted or an incorrect Nodepool ocid is passed) and hence we continuously trigger forceRefresh (when CA triggers Refresh) and therefore CA will indefinitely make GetNodePool calls to CP.
Introducing a max retry count and setting a delay to the next force refresh when the retries is exhausted will help prevent this.

Does this PR introduce a user-facing change?

No

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

None

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go [hunk 1]
```diff
@@ -249,11 +250,15 @@ func (m *ociManagerImpl) TaintToPreventFurtherSchedulingOnRestart(nodes []*apiv1
 }
 
 func (m *ociManagerImpl) forceRefresh() error {
-	err := m.nodePoolCache.rebuild(m.staticNodePools)
+	httpStatusCode, err := m.nodePoolCache.rebuild(m.staticNodePools, maxGetNodepoolRetries)
 	if err != nil {
+		if httpStatusCode == 404 {
```
Comment (trungng92) [inline cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:255]: If the OCI SDK returns a 404, is that considered an error?

File: cluster-autoscaler/cloudprovider/oci/nodepools/cache.go [hunk 2]
```diff
@@ -40,24 +40,33 @@ func (c *nodePoolCache) nodePools() map[string]*oke.NodePool {
 	return result
 }
 
-func (c *nodePoolCache) rebuild(staticNodePools map[string]NodePool) error {
+func (c *nodePoolCache) rebuild(staticNodePools map[string]NodePool, maxGetNodepoolRetries int) (httpStatusCode int, err error) {
 	klog.Infof("rebuilding cache")
+	var resp oke.GetNodePoolResponse
+	var statusCode int
 	for id := range staticNodePools {
```
Comment (trungng92) [inline cluster-autoscaler/cloudprovider/oci/nodepools/cache.go:47]: One consideration. We're looping over all of the node pools, but we exit/error out on the first one that errors out, which means you could have 20 node pools, and if the first one is deleted/errors out, then we won't reconcile any of them.

This is how the previous behavior worked as well, so we don't need to address it now, but we should note it as a bug (perhaps as a code comment and/or an internal ticket).

Comment (vbhargav875) [reply cluster-autoscaler/cloudprovider/oci/nodepools/cache.go:47]: Agreed

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go (hunk 1 above)
Comment (vbhargav875) [reply cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:255]: Yes, in case of 404, err is not nil.

**Comment by trungng92** [comment]
Changes look good to me

**Review by jlamillan — APPROVED** [review]
/lgtm
/approve

**Pull request merged**